src/FsAutoComplete/LspServers/AdaptiveFSharpLspServer.fs

```diff
@@ -3993,11 +4006,226 @@ type AdaptiveFSharpLspServer
+            |> Seq.toArray
+            |> Array.map locationToCallHierarchyItem
+            |> Async.parallel75
+            |> Async.map (Array.choose id)
+
+          return Some references
+        with e ->
+          trace |> Tracing.recordException e
+
+          logger.error (
+            Log.setMessage "CallHierarchyIncomingCalls Request Errored {p}"
+            >> Log.addContextDestructured "p" p
+            >> Log.addExn e
+          )
+
+          return! returnException e
+
+      }
+
+    override x.CallHierarchyOutgoingCalls p =
+      asyncResult {
+        let tags = [ "CallHierarchyOutgoingCalls", box p ]
+        use trace = fsacActivitySource.StartActivityForType(thisType, tags = tags)
+
+        try
+          logger.info (
+            Log.setMessage "CallHierarchyOutgoingCalls Request: {parms}"
+            >> Log.addContextDestructured "parms" p
+          )
+
+          let filePath = Path.FileUriToLocalPath p.Item.Uri |> Utils.normalizePath
+          let pos = protocolPosToPos p.Item.SelectionRange.Start
+          let! volatileFile = forceFindOpenFileOrRead filePath |> AsyncResult.ofStringErr
+          let! lineStr = tryGetLineStr pos volatileFile.Source |> Result.ofStringErr
+          and! tyRes = forceGetTypeCheckResults filePath |> AsyncResult.ofStringErr
+
+          let! implFile =
+            tyRes.GetCheckResults.ImplementationFile
+            |> Result.ofOption (fun () -> "No implementation file found")
+            |> Result.ofStringErr
+
+          let membersOrFunctions = ResizeArray<_>()
+          CallHierarchy.gatherFSharpMemberOrFunction (membersOrFunctions.Add) implFile.Declarations
+
+          let createOutGoingCall (range: FSharp.Compiler.Text.Range, x: FSharpMemberOrFunctionOrValue) =
+
+            let file = range.FileName |> Utils.normalizePath |> Path.LocalPathToUri
+            // let! tyRes = forceGetTypeCheckResults filePath |> AsyncResult.ofStringErr
+            // and! lineStr = tryGetLineStr pos volatileFile.Source |> Result.ofStringErr
+            let declRange =
+              try
+                fcsRangeToLsp x.DeclarationLocation
+              with e ->
+                LspRange.Zero
+
+            { To =
+                { Name = x.DisplayName
+                  Kind = SymbolKind.Function
+                  Tags = None
+                  Detail = None
+                  Uri = file
+                  Range = declRange
+                  SelectionRange = declRange
+                  Data = None }
+              FromRanges = [| fcsRangeToLsp range |] }
+
+
+          let allRemainingExpr =
+            membersOrFunctions
+            |> Seq.filter (fun (range, x) ->
+              let lookingFor =
+
+                x.IsFunction
+                || x.IsMethod
+                || x.IsPropertyGetterMethod
+                || x.IsPropertySetterMethod
+              // TODO get bodyRange and check if it is in the range
+              range.Start.Line >= pos.Line && lookingFor
```

Comment:
Are you trying to emulate visibility here? So checking for other items in this file that are after the trigger position that are one of these types? I'm having a hard time wrapping my head around the 'direction' of the incoming/outgoing part so I'm very slow here :)

As it is, it kinda feels like this filter could be pushed down into the traversal function you have, so that we don't create/allocate so much.

Reply:
Anything in the Outgoing implementation was some fever dream coding at like 11pm so I don’t know.